Remotion/Data/DomainObjects.UnitTests/Mapping/Builder/ClassDefinitionBuilderNodeTest.cs

```diff
@@ -0,0 +1,279 @@
+
+      mappingObjectFactoryMock.Verify();
+    }
+
+    [Test]
+    public void BeginBuildTypeDefinition_WithNonLeafNode_DoesNothing ()
+    {
+      var mappingObjectFactoryMock = new Mock<IMappingObjectFactory>();
+
+      var baseNode = new ClassDefinitionBuilderNode(typeof(Base), null);
+      var subNode = new ClassDefinitionBuilderNode(typeof(SubWithStorageGroup), baseNode);
+
+      baseNode.BeginBuildTypeDefinition(mappingObjectFactoryMock.Object);
+
+      Assert.That(baseNode.ClassDefinition, Is.Null);
+      Assert.That(subNode.ClassDefinition, Is.Null);
+    }
+
+    [Test]
+    public void GetBuiltTypeDefinition ()
+    {
+      var baseClassDefinition = ClassDefinitionObjectMother.CreateClassDefinition();
+      var mappingObjectFactoryMock = new Mock<IMappingObjectFactory>(MockBehavior.Strict);
+
+      mappingObjectFactoryMock.Setup(_ => _.CreateClassDefinition(typeof(Base), null)).Returns(baseClassDefinition);
+
+      var baseNode = new ClassDefinitionBuilderNode(typeof(Base), null);
+      baseNode.BeginBuildTypeDefinition(mappingObjectFactoryMock.Object);
+      baseNode.EndBuildTypeDefinition();
+
+      Assert.That(baseNode.GetBuiltTypeDefinition(), Is.SameAs(baseClassDefinition));
+    }
+
+    [Test]
+    public void GetBuiltTypeDefinition_NonReadOnly_Throws ()
+    {
+      var baseClassDefinition = ClassDefinitionObjectMother.CreateClassDefinition();
+      var mappingObjectFactoryMock = new Mock<IMappingObjectFactory>(MockBehavior.Strict);
+
+      mappingObjectFactoryMock.Setup(_ => _.CreateClassDefinition(typeof(Base), null)).Returns(baseClassDefinition);
+
+      var baseNode = new ClassDefinitionBuilderNode(typeof(Base), null);
+      baseNode.BeginBuildTypeDefinition(mappingObjectFactoryMock.Object);
+
+      Assert.That(
+          () => baseNode.GetBuiltTypeDefinition(),
+          Throws.InvalidOperationException
+              .With.Message.EqualTo("Cannot get the built type definition from a non read-only builder node."));
+    }
+
+    [Test]
+    public void AddDerivedClass_ReadOnlyBaseClass_Throws ()
+    {
+      var baseNode = new ClassDefinitionBuilderNode(typeof(Base), null);
+      baseNode.EndBuildTypeDefinition();
+
+      Assert.That(
+          () => new ClassDefinitionBuilderNode(typeof(Sub), baseNode),
+          Throws.ArgumentException
+              .With.ArgumentExceptionMessageEqualTo("The specified base class must not be read-only.", "baseClass"));
+    }
+
+    [Test]
+    public void AddDerivedClass_AlreadyAddedDerivedClass_Throws ()
+    {
+      var baseNode = new ClassDefinitionBuilderNode(typeof(Base), null);
+      var subNode = new ClassDefinitionBuilderNode(typeof(Sub), baseNode);
+
+      Assert.That(
+          () => PrivateInvoke.InvokeNonPublicMethod(baseNode, "AddDerivedClass", subNode),
+          Throws.ArgumentException
+              .With.ArgumentExceptionMessageEqualTo("Cannot add the specified derived class as it is already a child.", "derivedClass"));
```

Comment:
does this imply we can't construct a public scenario to trigger this exception? since I already requested to change it to an assertion, this might further indicate this test isn't meaningful. If there is no way to construct code for this assertion, we might even consider making it a debug-assertion,